### PR TITLE
Support header extension removal

### PR DESCRIPTION
**Pull request by JonathanLennox**
Allow removal of RTP header extension elements; cache changes to header extensions until explicitly encoded.

### CODE REVIEW

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt [hunk 1]
```diff
@@ -173,17 +174,62 @@ open class RtpPacket(
...
+        encodedHeaderExtensions.forEach { ext ->
+            if (ext.id == extensionId) {
+                return ext
+            }
+        }
+        return null
+    }
+
+    val hasExtensions: Boolean
+        get() = pendingHeaderExtensions?.isNotEmpty() ?: hasEncodedExtensions
```
Comment (bgrozev) [inline src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:191]: If `pendingHeaderExtensions` is empty (but not null), this will return `false` regardless of the value of `hasEncodedExtensions`. Is this the intent?

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt [hunk 2]
```diff
@@ -173,17 +174,62 @@ open class RtpPacket(
...
             }
         }
         return null
     }
 
+    private fun createPendingHeaderExtensions(removeIf: ((HeaderExtension) -> Boolean)?) {
+        if (pendingHeaderExtensions != null) {
+            return
+        }
+        pendingHeaderExtensions = ArrayList<HeaderExtension>().also { l: ArrayList<HeaderExtension> ->
```
Comment (bgrozev) [inline src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:208]: I don't think the type of `l` needs to be explicit

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt [hunk 3]
```diff
@@ -208,28 +269,33 @@ open class RtpPacket(
...
-                extDataLength +
-                3 /* padding */
+        val baseHeaderLength = RtpHeader.FIXED_HEADER_SIZE_BYTES + csrcCount * 4
+
+        val newExtHeaderLength = if (pendingHeaderExtensions.isEmpty()) {
+            0
+        } else {
+            val rawHeaderLength = RtpHeader.EXT_HEADER_SIZE_BYTES +
+                pendingHeaderExtensions.sumBy { h -> h.totalLengthBytes }
+            val paddingLength = (4 - (rawHeaderLength % 4)) % 4
```
Comment (bgrozev) [inline src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:288]: We have `RtpUtils.getNumPaddingBytes` which does the same (I don't know if the performance is different)

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt [hunk 4]
```diff
@@ -242,136 +308,81 @@ open class RtpPacket(
...
+        this.pendingHeaderExtensions = null
     }
 
     /**
      * Return the total length of the extensions in this packet, including the extension header
      */
     private val extensionBlockLength: Int
         get() {
-            if (!hasExtensions) {
+            if (!hasEncodedExtensions) {
```
Comment (bgrozev) [inline src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:363]: It might be cleaner to inline this function (it's only used once and doesn't need the `hasEncodedExtensions` check).

File: src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt [hunk 5]
```diff
@@ -303,5 +316,157 @@ class RtpPacketTest : ShouldSpec() {
...
+                    ext shouldNotBe null
+                    ext as RtpPacket.HeaderExtension
+                    ext.id shouldBe 1
+                    ext.dataLengthBytes shouldBe 2
+                    ext.currExtBuffer.getShort(ext.currExtOffset + 1) shouldBe 0xFFFF.toShort()
+
+                    val ext2 = rtpPacket.getHeaderExtension(2)
+                    ext2 shouldBe null
+
+                    // The offset is the start of the ext, add 1 to move past the header to get the data
```
Comment (bgrozev) [inline src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt:373]: This comment doesn't seem related

File: src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt [hunk 6]
```diff
@@ -303,5 +316,157 @@ class RtpPacketTest : ShouldSpec() {
...
+                    rtpPacket.hasExtensions shouldBe false
+                    rtpPacket should haveSamePayload(rtpPacketNoExtensions)
+                }
+            }
+        }
+        context("Removing all but a given header extension") {
+            context("when a header extension remains") {
+                val rtpPacket = rtpPacketWithExtensionsWithPaddingBetween.clone()
+                rtpPacket.removeHeaderExtensionsExcept(Collections.singleton(1))
+                rtpPacket.encodeHeaderExtensions()
```
Comment (bgrozev) [inline src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt:358]: Is there an easy way to run this test (and the other similar tests) with and without applying `encodeHeaderExtensions()`?

**Review by bgrozev — APPROVED** [review]

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt [hunk 7]
```diff
@@ -173,17 +175,77 @@ open class RtpPacket(
...
+        pendingHeaderExtensions?.removeIf { h -> h.id == id }
+            ?: createPendingHeaderExtensions { h -> h.id == id }
+    }
+
+    /**
+     * Removes all header extensions except those with ID values in [retain]
+     */
+    fun removeHeaderExtensionsExcept(retain: Set<Int>) {
+        pendingHeaderExtensions?.removeIf { h -> !retain.contains(h.id) }
+            ?: createPendingHeaderExtensions { h -> !retain.contains(h.id) }
```
Comment (bgrozev) [inline src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:246]: This may re-encode the extensions needlessly if `retain` contains all encoded extensions, and this may be ea common case if we use this to clear non-signalled extensions in jmt. Should it be optimized to reset `pendingHeaderExtensions` to `null` if `pendingExtensions.length == encodedExtensions.length`?

File: src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt (hunk 5 above)
Comment (JonathanLennox) [reply src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt:373]: Moved the code without moving the comment.

File: src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt (hunk 6 above)
Comment (JonathanLennox) [reply src/test/kotlin/org/jitsi/rtp/rtp/RtpPacketTest.kt:358]: Hm, you mean to get the pending headers?  I'm not sure how easy it would be without duplicating the tests.

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt (hunk 7 above)
Comment (JonathanLennox) [reply src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:246]: Well, I'm planning to use it to remove all hop-by-hop extensions in jmt (see the jmt PR), so in the normal cases we'll have removed something (i.e. tcc or abs-send-time).

`encodedExtensions` is an `Iterator<HeaderExtension>` not a `Collection` so it won't have a `length`, but a `removedSomething` flag is certainly doable.

**Review by bgrozev — DISMISSED** [review]

File: src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt (hunk 7 above)
Comment (JonathanLennox) [reply src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:246]: Hm, I tried writing this, but the issue is that if you did a `removeHeaderExtensionsExcept` which ends up being a no-op, followed by an `addHeaderExtension`, you'd end up creating the pending header extension vector twice.

Since for now I'd anticipate that we'd always remove a hop-by-hop header (a tcc or abs-send-time), I think not doing this optimization is fine.

Comment (bgrozev) [reply src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt:246]: Okay. I was going to suggest swapping the extension stripper and TCC tagger to avoid re-encoding, but I guess we can and should optimize only if it proves necessary.

**Review by bgrozev — APPROVED** [review]

**Review by bgrozev — APPROVED** [review]
